### PR TITLE
CASSANDRASC-99 Break restore job into stage and import phases and persist restore sl…

### PR DESCRIPTION
**Pull request by yifan-c**
…ice status on phase completion

### CODE REVIEW

File: src/main/java/org/apache/cassandra/sidecar/db/RestoreJob.java [hunk 1]
```diff
@@ -156,13 +123,28 @@ private RestoreJob(Builder builder)
...
 
+    /**
+     * When a job object is materialized from database, and it has non-null consistencyLevel, the range of the slices
+     * of the restore is managed by Sidecar server, meaning that server should assign slices to sidecar instances and
+     * check whether the job has met the consistency level to complete the job; otherwise, sidecar instances are just
+     * simple workers and rely on client for decision-making.
+     * Depending on the return value, call-sites select the correct handling for the restore jobs and slices
+     * @return true when sidecar server manages the slices based on the token range; otherwise, return false.
+     */
+    public boolean isRangeManagedByServer()
```
Comment (yifan-c) [inline src/main/java/org/apache/cassandra/sidecar/db/RestoreJob.java]: naming is hard. Open to suggestions

File: src/main/java/org/apache/cassandra/sidecar/config/RestoreJobConfiguration.java [hunk 2]
```diff
@@ -48,4 +48,11 @@ public interface RestoreJobConfiguration
      * @return time to live for restore job tables: restore_job and restore_slice
      */
     long restoreJobTablesTtlSeconds();
+
+    /**
+     * Controls whether the feature to allow server-side the slice assignment and verification based on the token range
+     * is enabled or not.
+     * @return true if enabled; false otherwise
+     */
+    boolean isServerManagedRangeEnabled();
```
Comment (yifan-c) [inline src/main/java/org/apache/cassandra/sidecar/config/RestoreJobConfiguration.java]: Naming is hard. Open to suggestions.

File: common/src/main/java/org/apache/cassandra/sidecar/common/data/CreateRestoreJobRequestPayload.java [hunk 3]
```diff
@@ -43,6 +44,7 @@ public class CreateRestoreJobRequestPayload
     private final RestoreJobSecrets secrets;
     private final SSTableImportOptions importOptions;
     private final long expireAtInMillis;
+    private final String consistencyLevel; // Nullable / optional field
```
Comment (frankgh) [inline common/src/main/java/org/apache/cassandra/sidecar/common/data/CreateRestoreJobRequestPayload.java]: we can use annotations if we want to mark them as nullable? no?

File: common/src/main/java/org/apache/cassandra/sidecar/common/data/RestoreSliceStatus.java [hunk 4]
```diff
@@ -18,15 +18,44 @@
...
     FAILED,
-    ABORTED
+    ABORTED,
+    COMMITTING(SUCCEEDED, FAILED, ABORTED),
+    STAGED(COMMITTING, FAILED, ABORTED),
+    PROCESSING(STAGED, FAILED, ABORTED),
+    EMPTY(PROCESSING, FAILED, ABORTED);
+
+    // Do not use EnumSet, since validTargetStatuses is assigned on constructing and enums are not available yet.
+    private final Set<RestoreSliceStatus> validTargetStatuses;
```
Comment (frankgh) [inline common/src/main/java/org/apache/cassandra/sidecar/common/data/RestoreSliceStatus.java]: NIT
```suggestion
    private final Set<RestoreSliceStatus> validTargetStatusSet;
```

File: common/src/test/java/org/apache/cassandra/sidecar/common/data/RestoreSliceStatusTest.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import static org.apache.cassandra.sidecar.common.data.RestoreSliceStatus.STAGED;
+import static org.apache.cassandra.sidecar.common.data.RestoreSliceStatus.SUCCEEDED;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class RestoreSliceStatusTest
+{
+    @Test
+    void testStatusAdvancing()
+    {
+        EMPTY.advanceTo(PROCESSING);
```
Comment (frankgh) [inline common/src/test/java/org/apache/cassandra/sidecar/common/data/RestoreSliceStatusTest.java]: maybe assert?
```suggestion
        assertThat(EMPTY.advanceTo(PROCESSING)).isEqualTo(PROCESSING);
```

File: src/main/java/org/apache/cassandra/sidecar/config/RestoreJobConfiguration.java [hunk 6]
```diff
@@ -48,4 +48,11 @@ public interface RestoreJobConfiguration
      * @return time to live for restore job tables: restore_job and restore_slice
      */
     long restoreJobTablesTtlSeconds();
+
+    /**
+     * Controls whether the feature to allow server-side the slice assignment and verification based on the token range
```
Comment (frankgh) [inline src/main/java/org/apache/cassandra/sidecar/config/RestoreJobConfiguration.java]: I'm not following the description of this javadoc. Can you elaborate what we are trying to achieve here?

**Review by frankgh — COMMENTED** [review]
Looks good! I've left some minor comments

File: src/main/java/org/apache/cassandra/sidecar/config/RestoreJobConfiguration.java (hunk 6 above)
Comment (yifan-c) [reply src/main/java/org/apache/cassandra/sidecar/config/RestoreJobConfiguration.java]: The goal is to having a boolean flag to toggle the "server-side the slice assignment and verification based on the token range". Server side slice assignment is to select and download the slices based on its owning token range. Verification is to check whether the slices imported can satisfy the specified consistency level.

File: src/main/java/org/apache/cassandra/sidecar/db/RestoreJob.java [hunk 7]
```diff
@@ -76,10 +75,13 @@ public static RestoreJob from(@NotNull Row row) throws DataObjectMappingExceptio
                .jobSecrets(decodeJobSecrets(row.getBytes("blob_secrets")))
                .expireAt(row.getTimestamp("expire_at"))
                .sstableImportOptions(decodeSSTableImportOptions(row.getBytes("import_options")));
+        String consistencyLevel = row.getString("consistency_level");
+        builder.manager(consistencyLevel == null ? Manager.SPARK : Manager.SIDECAR)
```
Comment (JeetKunDoug) [inline src/main/java/org/apache/cassandra/sidecar/db/RestoreJob.java]: Took me a moment to figure out why consistencyLevel would be null if the manager is SPARK - can we document why this is the case here please?

**Review by frankgh — APPROVED** [review]
+1 the patch looks good

**Pull request closed without merging**